### PR TITLE
Gml 1284 object oriented schema

### PR DESCRIPTION
**Pull request by parkererickson-tg**

### CODE REVIEW

**Review by qe-tigergraph — COMMENTED** [review]
Unit Test: SUCCESS, e2e Test: SUCCESS, Jenkins_job:http://192.168.99.101:30080/job/mlwb_build/829/

**Review by qe-tigergraph — APPROVED** [review]
QE Approved

File: tests/test_OGM.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+class TestHomogeneousOGM(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.conn = make_connection(graphname="Cora")
+
+    def test_init(self):
+        g = Graph(self.conn)
+        attrs = g.vertex_types["Paper"].attributes
+        self.assertIn(attrs.keys(), "x")
```
Comment (billshitg) [inline tests/test_OGM.py:17]: Is this checking `attrs.keys() in "x"`?

File: pyTigerGraph/schema.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        val_type = _py_to_tg_type(attr_type.__args__[1])
+        if key_type.lower() in MAP_KEY_TYPES:
+            if val_type.lower() in COLLECTION_VALUE_TYPES:
+                return "MAP<"+key_type+","+val_type+">"
+            else:
+                raise TigerGraphException(val_type + " not a valid type for the value type in MAPs.")
+        else:
+            raise TigerGraphException(key_type + " not a valid type for the key type in MAPs.")
+    else:
+        print(attr_type)
```
Comment (billshitg) [inline pyTigerGraph/schema.py:76]: Is this intended for debugging or actually in prod?

File: pyTigerGraph/schema.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+        self.primary_id_type = self.__annotations__[primary_id]
+        if not(_py_to_tg_type(self.primary_id_type).lower() in PRIMARY_ID_TYPES):
+            raise TigerGraphException(self.primary_id_type, "is not a supported type for primary IDs.")
+
+    @classmethod
+    def add_attribute(self, attribute, attribute_type):
+        for attr in self.attributes.values():
+            if attr == attribute:
+                raise TigerGraphException(attribute + " already exists as an attribute on "+self.vertex_type + " vertices")
+        self.attribute_edits["ADD"].append(attribute)
```
Comment (billshitg) [inline pyTigerGraph/schema.py:104]: Is attribute_type not used in this function?

File: pyTigerGraph/schema.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        return self.edge_type
+
+class Graph():
+    def __init__(self, conn=None):
+        self._vertex_types = {}
+        self._edge_types = {}
+        self._vertex_edits = {"ADD": {}, "DELETE": {}}
+        self._edge_edits = {"ADD": [], "DELETE": []}
+        if conn:
+            db_rep = conn.getSchema()
```
Comment (billshitg) [inline pyTigerGraph/schema.py:183]: Do we want force=True here?

File: pyTigerGraph/schema.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        attrs = edge.attributes
+        for attr in attrs.keys():
+            if attr == "from_vertex" or attr == "to_vertex" or attr == "is_directed" or attr == "reverse_edge":
+                continue
+            else:
+                gsql_def += ", "
+                gsql_def += attr + " "+_py_to_tg_type(attrs[attr])
+        gsql_def += ")"
+
+        print(gsql_def)
```
Comment (billshitg) [inline pyTigerGraph/schema.py:258]: Is this intended for debugging or prod? Given the same comment above, I am wondering do we want a "show gsql" function or something like that?

File: pyTigerGraph/schema.py (hunk 2 above)
Comment (parkererickson-tg) [reply pyTigerGraph/schema.py:76]: Whoops, yeah debug

File: pyTigerGraph/schema.py (hunk 4 above)
Comment (parkererickson-tg) [reply pyTigerGraph/schema.py:183]: Yep

File: pyTigerGraph/schema.py (hunk 5 above)
Comment (parkererickson-tg) [reply pyTigerGraph/schema.py:258]: This is debugging, I was going to edit this in future stories, as the creation/deletion of types is in the backlog. Got ahead of myself here...

File: pyTigerGraph/schema.py (hunk 3 above)
Comment (parkererickson-tg) [reply pyTigerGraph/schema.py:103]: Removed, might need to re-instate it when I get to that logic

File: tests/test_OGM.py (hunk 1 above)
Comment (parkererickson-tg) [reply tests/test_OGM.py]: Yeah... fixed.

**Review by qe-tigergraph — COMMENTED** [review]
Unit Test: SUCCESS, e2e Test: SUCCESS, Jenkins_job:http://192.168.99.101:30080/job/mlwb_build/844/

**Review by qe-tigergraph — APPROVED** [review]
QE Approved

**Pull request merged**